### PR TITLE
Fix etag propagation for single file shares

### PR DESCRIPTION
**Pull request by icewind1991**
To test:
- create 2 users user1 and user2
- create a text file as user1 and share it with user2
- check the root etag for user2
- update the text file as user1
- check if the root etag for user2 has been changed

cc @DeepDiver1975 @PVince81

### CODE REVIEW

**Comment by icewind1991** [comment]
8.1? breaks sync

**Comment by PVince81** [comment]
Wow, infinite loop!

@icewind1991 please have a look

**Comment by icewind1991** [comment]
Added a unit test and changed to approach to prevent the infinite loop

File: apps/files_sharing/lib/propagation/recipientpropagator.php [hunk 1]
```diff
@@ -101,18 +108,24 @@ public function markDirty($share, $time = null) {
...
-				$user = $share['uid_owner'];
+	public function propagateById($id) {
+		$shares = Share::getAllSharesForFileId($id);
+		foreach ($shares as $share) {
+			// propagate down the share tree
+			$this->markDirty($share, microtime(true));
+
+			// propagate up the share tree
+			$user = $share['uid_owner'];
+			if($user !== $this->userId) {
```
Comment (PVince81) [inline apps/files_sharing/lib/propagation/recipientpropagator.php]: Was this missing ? Or is it an additional optimization ?
Or required for this fix ?

**Comment by PVince81** [comment]
@icewind1991 mind explaning the change a little bit ? (the commit comment is a bit sparse)
Thanks.

**Comment by PVince81** [comment]
It is also unclear to me why single file share require special handling, and also I don't understand in what way your change makes it work.

**Comment by PVince81** [comment]
- [x] TEST: propagation user1->user2 with file in root: etag of root changed
- [x] TEST: propagation user2->user1 with file in root: etag of root changed
- [x] TEST: propagation user1->user2 with file in subdir: etag of root + subdir changed
- [x] TEST: propagation user2->user1 with file in subdir: etag of root + subdir changed

**Comment by icewind1991** [comment]
> It is also unclear to me why single file share require special handling, and also I don't understand in what way your change makes it work.

Before this etag propagation hooks into the change propagation (which is responsible for updating etag/mtime for parent folders) but the change propagator only emits hooks for the parent folders of changed files. This is nice with shared folders or more complex share setups but doesn't work if the file changed itself is shared

File: apps/files_sharing/lib/propagation/recipientpropagator.php (hunk 1 above)
Comment (icewind1991) [inline apps/files_sharing/lib/propagation/recipientpropagator.php]: required for this fix, also small optimization

File: apps/files_sharing/lib/propagation/changewatcher.php [hunk 2]
```diff
@@ -39,6 +46,10 @@ public function writeHook($params) {
 		if ($mount instanceof SharedMount) {
 			$this->propagateForOwner($mount->getShare(), $mount->getInternalPath($fullPath), $mount->getOwnerPropagator());
 		}
+		$info = $this->baseView->getFileInfo($path);
+		if ($info) {
+			$this->recipientPropagator->propagateById($info->getId());
```
Comment (PVince81) [inline apps/files_sharing/lib/propagation/changewatcher.php]: Is that the part that detects if this is a file share ?
Will the change also affect folder shares ?

Comment (icewind1991) [inline apps/files_sharing/lib/propagation/changewatcher.php]: only if the subject of the write hook has a share entry.

So only for single file shares

Comment (PVince81) [inline apps/files_sharing/lib/propagation/changewatcher.php]: Mind adding a small inline comment for that ?
Thanks :smile:

**Comment by PVince81** [comment]
:+1:

**Comment by icewind1991** [comment]
@owncloud-bot retest this please

**Comment by DeepDiver1975** [comment]
@icewind1991 please rebase for oracle ci - thx

File: apps/files_sharing/lib/propagation/changewatcher.php (hunk 2 above)
Comment (MorrisJobke) [inline apps/files_sharing/lib/propagation/changewatcher.php]: @icewind1991 Please add the comment

Comment (icewind1991) [inline apps/files_sharing/lib/propagation/changewatcher.php]: doner

**Comment by PVince81** [comment]
@schiesbn @nickvergessen @DeepDiver1975 second reviewer please

**Comment by DeepDiver1975** [comment]
:+1:

**Comment by PVince81** [comment]
Regression: https://github.com/owncloud/core/issues/18034

**Pull request merged**